packages/flutter/lib/src/services/autofill.dart

```diff
@@ -0,0 +1,280 @@
+  ///
+  /// Returns null if there's no matching [AutofillClient].
+  AutofillClient getAutofillClient(String uniqueIdentifier);
+
+  /// The collection of [AutofillClient]s currently tied to this [AutofillScope].
+  ///
+  /// The [AutofillClient]s should appear in a sensible order, as the autofill
+  /// service will see these [AutofillClient]s in the exact same order.
+  Iterable<AutofillClient> get autofillClients;
+
+  /// Allows an [AutofillTrigger] to attach to this scope. This method should be
+  /// called in lieu of [TextInput.attach], when the [AutofillTrigger] wishes to
+  /// participate in autofill.
+  TextInputConnection attach(AutofillTrigger trigger, TextInputConfiguration configuration);
+}
+
+@immutable
+class _AutofillScopeTextInputConfiguration extends TextInputConfiguration {
+  _AutofillScopeTextInputConfiguration({
+    @required this.allConfigurations,
+    @required TextInputConfiguration currentClientConfiguration,
+  }) : assert(allConfigurations != null),
+       assert(currentClientConfiguration != null),
+       super(inputType: currentClientConfiguration.inputType,
+         obscureText: currentClientConfiguration.obscureText,
+         autocorrect: currentClientConfiguration.autocorrect,
+         smartDashesType: currentClientConfiguration.smartDashesType,
+         smartQuotesType: currentClientConfiguration.smartQuotesType,
+         enableSuggestions: currentClientConfiguration.enableSuggestions,
+         inputAction: currentClientConfiguration.inputAction,
+         textCapitalization: currentClientConfiguration.textCapitalization,
+         keyboardAppearance: currentClientConfiguration.keyboardAppearance,
+         actionLabel: currentClientConfiguration.actionLabel,
+         autofillConfiguration: currentClientConfiguration.autofillConfiguration,
+       );
+
+  final Iterable<TextInputConfiguration> allConfigurations;
+
+  @override
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> result = super.toJson();
+    result['allFields'] = allConfigurations
+      .map((TextInputConfiguration configuration) => configuration.toJson())
+      .toList(growable: false);
+    return result;
+  }
+}
+
+/// A partial implementation of [AutofillScope].
+///
+/// The mixin provides a default implementation for [AutofillScope.attach], which,
+/// when called, caches the list of [AutofillClient]s by storing the contents of
+/// [autofillClients] in a [Map]. This allows the implementation of [getAutofillClient]
+/// to be more efficient.
+mixin AutofillScopeMixin implements AutofillScope {
+  final Map<String, AutofillClient> _clients = <String, AutofillClient>{};
+
+  @override
+  AutofillClient getAutofillClient(String tag) => _clients[tag];
+
+  @override
+  TextInputConnection attach(AutofillTrigger trigger, TextInputConfiguration configuration) {
+    assert(trigger != null);
+    // Caches clients on attach.
+    _clients.clear();
+    for (final AutofillClient c in autofillClients)
```

Comment:
We usually don't use abbreviations like "c". `for (final AutofillClient client in autofillClients)` would be OK.